munkipkg

```diff
@@ -619,6 +659,157 @@ def build_distribution_pkg(build_info, options):
+        display("SUCCESS " + output['success-message'], options.quiet, "altool")
+    except KeyError:
+        raise MunkiPkgError("Unexpected output from altool")
+
+    return request_uuid
+
+
+def get_notarization_state(request_uuid, build_info, options):
+    '''Checks for result of notarization process'''
+    state = {}
+    cmd = [
+        XCRUN,
+        'altool',
+        '--notarization-info',
+        request_uuid,
+        '--username',
+        build_info['notarization_info']['username'],
+        '--password',
+        build_info['notarization_info']['password'],
+        '--output-format',
+        'xml',
+    ]
+    retcode, proc_stdout, proc_stderr = run_subprocess(cmd)
+
+    try:
+        output = readPlistFromString(proc_stdout)
+    except ExpatError:
+        print(proc_stderr, file=sys.stderr)
+        raise MunkiPkgError("Notarization check failed. Unable to run xcrun altool")
+
+    if retcode or 'notarization-info' not in output:
+        print("altool: " + output.get('success-message', 'Unexpected response'))
+    else:
+        state['log_url'] = output['notarization-info'].get('LogFileURL', '')
+        state['status'] = output['notarization-info'].get('Status', 'Unknown')
+        state['code'] = output['notarization-info'].get('Status Code', None)
+        state['message'] = output['notarization-info'].get('Status Message', '')
+        return state
+
+
+def notarization_done(state, sleep_time, options):
+    '''Evaluates whether notarization is still in progress'''
+    if state['status'] == 'success':
+        display("Notarization successful. {}".format(state['message']), options.quiet)
+        return True
+    elif state['status'] == 'in progress':
+        display(
+            "Notarization still in progress. Trying again in {} seconds".format(
+                sleep_time
+            ),
+            options.quiet,
+        )
+        return False
+    else:
+        display(
+            "Notarization unsuccessful:\n"
+            "\tStatus: {}\n"
+            "\tStatus Code: {}\n"
+            "\tStatus Message: {}\n"
+            "\tLogFileURL: {}".format(
+                state['status'], state['code'], state['message'], state['log_url']
+            ),
+            options.quiet,
+        )
+        raise MunkiPkgError("Notarization failed")
+
+
+def wait_for_notarization(request_uuid, build_info, options):
+    '''Checks notarization state until it is done or we exceed the timeout value'''
+    display("Getting notarization state", options.quiet)
+    timeout = build_info['notarization_info'].get('staple_timeout', STAPLE_TIMEOUT)
+    counter = 0
+    sleep_time = STAPLE_SLEEP
+
+    while counter < timeout:
+        time.sleep(sleep_time)
+        state = get_notarization_state(request_uuid, build_info, options)
+        counter += sleep_time
+        sleep_time += STAPLE_SLEEP
+
+        if notarization_done(state, sleep_time, options):
+            break
+
```

Comment:
Looks like it will exit "normally" after the staple_timeout -- in other words, when wait_for_notarization returns notarization might have been successful, or we might have just timed out. So in lines 880 and 881 we might proceed to staple even if the notarization is not complete.

I think it would be better to recognize the timeout and 1) Not try to staple, and 2) Print out information that would allow the admin to manually staple later if/when they get notification of a successful notarization.

Reply:
I agree. Definitely bug I created.

See the fixup 5af2ae3.

btw I do nothing with exit codes. munkipkg exits with 0 when notaratization fails or times out. I was thinking primary job of munkipkg is to build packages. If package is built correctly process should exit with code 0 even when notarization failed. Do you agree?

Reply:
Not sure I do. Primary job might be to build pkgs, but if you give it extra jobs it should tell if if/when it fails.

Reply:
See 1b13928.

In case of problem or failed notarization we return with -1 from `build()` and it gets passed to `sys.exit()`.
When we give up on waiting for notarization we still exit with 0.